app/code/Magento/InventorySales/Observer/CatalogInventory/CancelOrderItemObserver.php

```diff
@@ -91,61 +94,140 @@ public function __construct(
      */
     public function execute(EventObserver $observer)
     {
-        /** @var OrderItem $item */
-        $item = $observer->getEvent()->getItem();
-        $qty = $item->getQtyToCancel();
-        if ($this->canCancelOrderItem($item) && $qty) {
-            try {
-                $productSku = $this->getSkusByProductIds->execute(
-                    [$item->getProductId()]
-                )[$item->getProductId()];
-            } catch (NoSuchEntityException $e) {
-                /**
-                 * As it was decided the Inventory should not use data constraints depending on Catalog
-                 * (these two systems are not highly coupled, i.e. Magento does not sync data between them, so that
-                 * it's possible that SKU exists in Catalog, but does not exist in Inventory and vice versa)
-                 * it is necessary for Magento to have an ability to process placed orders even with
-                 * deleted or non-existing products
-                 */
-                return;
+        /** @var OrderItemInterface $orderItem */
+        $orderItem = $observer->getEvent()->getItem();
+        if ($this->canCancelOrderItem($orderItem)) {
```

Comment:
There are a couple of nested code blocks that maybe can be avoided with early exit conditions.
What I propose is something like this:
```
    public function execute(EventObserver $observer)
    {
        /** @var OrderItemInterface $orderItem */
        $orderItem = $observer->getEvent()->getItem();

        if (!$this->canCancelOrderItem($orderItem)) {
            return;
        }

        $itemsToSell = $this->getItemsToSell($orderItem);
        if (!empty($itemsToSell)) {
            return;
        }

        $websiteId = $orderItem->getStore()->getWebsiteId();
        $websiteCode = $this->websiteRepository->getById($websiteId)->getCode();
        $salesChannel = $this->salesChannelFactory->create([
            'data' => [
                'type' => SalesChannelInterface::TYPE_WEBSITE,
                'code' => $websiteCode,
            ]
        ]);

        /** @var SalesEventInterface $salesEvent */
        $salesEvent = $this->salesEventFactory->create([
            'type' => SalesEventInterface::EVENT_ORDER_CANCELED,
            'objectType' => SalesEventInterface::OBJECT_TYPE_ORDER,
            'objectId' => (string)$orderItem->getOrderId(),
        ]);

        $this->placeReservationsForSalesEvent->execute($itemsToSell, $salesChannel, $salesEvent);

        $this->priceIndexer->reindexRow($orderItem->getProductId());
}
```

Note: in my code, I assume that the `$this->priceIndexer->reindexRow($orderItem->getProductId());` should be executed only if the previous checks pass; I'm not sure if it should be always executed, in which case my proposal is not effective.

Let me know